lib/STM_Thread.ml

```diff
@@ -0,0 +1,39 @@
+open STM_Core
+
+module Make (Spec: Spec) = struct
+
+  open Util
+  open QCheck
+  open STM_Core.Make(Spec)
+
+  exception ThreadNotFinished
+
+  (* Parallel agreement property based on [Threads] *)
+  let agree_prop_thread (seq_pref,cmds1,cmds2) =
+    assume (all_interleavings_ok seq_pref cmds1 cmds2 Spec.init_state);
+    let sut = Spec.init_sut () in
+    let obs1,obs2 = ref (Error ThreadNotFinished), ref (Error ThreadNotFinished) in
+    let pref_obs = interp_sut_res sut seq_pref in
+    let wait = ref true in
+    let th1 = Thread.create (fun () -> while !wait do Thread.yield () done; obs1 := try Ok (interp_sut_res sut cmds1) with exn -> Error exn) () in
+    let th2 = Thread.create (fun () -> wait := false; obs2 := try Ok (interp_sut_res sut cmds2) with exn -> Error exn) () in
+    let () = Thread.join th1 in
+    let () = Thread.join th2 in
+    let obs1 = match !obs1 with Ok v -> v | Error exn -> raise exn in
+    let obs2 = match !obs2 with Ok v -> v | Error exn -> raise exn in
+    let ()   = Spec.cleanup sut in
+    check_obs pref_obs obs1 obs2 Spec.init_state
+      || Test.fail_reportf "  Results incompatible with linearized model\n\n%s"
+         @@ print_triple_vertical ~fig_indent:5 ~res_width:35
+           (fun (c,r) -> Printf.sprintf "%s : %s" (Spec.show_cmd c) (show_res r))
+           (pref_obs,obs1,obs2)
+
+  let agree_test_par ~count ~name =
```

Comment:
```ocaml
  let agree_prop_thread ...
  let agree_test_par ...
```
is a bit inconsistent with the `Domain` functor's naming.

Also, I've tried to keep the 'property' and 'test' identifiers aligned, so that they only differ in four characters (`prop` vs `test`) - but I've not written this design decision down anywhere up until now.